### ISSUE TITLE
Normalize and correct infix operator terminology

### ISSUE THREAD

**Issue by eulerfx** [SuaveIO/suave#361]
In `Http.fsi`:

```fsharp
  /// Compose (bind) two arguments, 'first' and 'second', so that the result of
  /// the composition can be applied to an argument of 'input' and then passed
  /// to 'second', if 'first' yields a value.
  val inline (>>=) : first:('T -> Async<'U option>) -> second:('U -> Async<'V option>) -> input:'T -> Async<'V option>

  /// Left-to-right Kleisli composition of monads.
  val inline (>=>) : first:('T -> 'U option) -> second:('T -> 'U option) -> input:'T -> 'U option

  /// Left-to-right Kleisli composition of web parts.
  val inline (<|>) : first:WebPart -> second:WebPart -> WebPart

```

The bind operator `(>>=)` usually has type (here specialized to Async): 

```fsharp
val (>>=) : Async<'a> -> ('a -> Async<'b>) -> Async<'b>
```
Similarly so in F# workflow builders (but order reversed).

And Kleisli composition `(>=>)` has type:

```fsharp
val (>=>) : ('a -> Async<'b>) -> ('b -> Async<'c>) -> ('a -> Async<'c>)
```

As seen in Suave, what is called the bind operator `(>>=)` is actually Kleisli composition, not bind. In turn, the operator `(>=>)` is erroneously called Kleisli composition. In fact, looking at the type of `(>=>)`, I'm not too sure what the meaning is, but I'm assuming that it tries the first, then second. I'm not sure what `(<|>)` means, but from the looks of it, it seems similar to `(>=>)`.

I don't think that F# operator naming and semantics *have* to match Haskell, but it seems appropriate to keep it consistent. More importantly, if using names like "Kleisli composition" they should be used correctly. Better yet, remove this terminology from the docs all together - F# already has a stigma of being complex and this isn't helping.

**Comment by ademar** [comment]
This is already addressed in our working branch - just waiting for @haf to
merge it.

On Thursday, 17 December 2015, Lev Gorodinski <notifications@github.com>
wrote:

> In Http.fsi:
>
>   /// Compose (bind) two arguments, 'first' and 'second', so that the result of
>   /// the composition can be applied to an argument of 'input' and then passed
>   /// to 'second', if 'first' yields a value.
>   val inline (>>=) : first:('T -> Async<'U option>) -> second:('U -> Async<'V option>) -> input:'T -> Async<'V option>
>
>   /// Left-to-right Kleisli composition of monads.
>   val inline (>=>) : first:('T -> 'U option) -> second:('T -> 'U option) -> input:'T -> 'U option
>
>   /// Left-to-right Kleisli composition of web parts.
>   val inline (<|>) : first:WebPart -> second:WebPart -> WebPart
>
> The bind operator (>>=) usually has type (here specialized to Async):
>
> val (>>=) : Async<'a> -> ('a -> Async<'b>) -> Async<'b>
>
> Similarly so in F# workflow builders (but order reversed).
>
> And Kleisli composition (>=>) has type:
>
> val (>=>) : ('a -> Async<'b>) -> ('b -> Async<'c>) -> ('a -> Async<'c>)
>
> As seen in Suave, what is called the bind operator (>>=) is actually
> Kleisli composition, not bind. In turn, the operator (>=>) is erroneously
> called Kleisli composition. In fact, looking at the type of (>=>), I'm
> not too sure what the meaning is, but I'm assuming that it tries the first,
> then second. I'm not sure what (<|>) means, but from the looks of it, it
> seems similar to (>=>).
>
> I don't think that F# operator naming and semantics *have* to match
> Haskell, but it seems appropriate to keep it consistent. More importantly,
> if using names like "Kleisli composition" they should be used correctly.
> Better yet, remove this terminology from the docs all together - F# already
> has a stigma of being complex and this isn't helping.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SuaveIO/suave/issues/361>.
>


-- 
Ademar Gonzalez
+1-647-891-3606
http://ademar.name
https://github.com/ademar

**Comment by eulerfx** [comment]
OK, cool! 

I think there is a similar issue with the use of applicatives. I don't think functions of type:

```fsharp
HttpContext -> Async<HttpContext option>
```
are best characterized as applicatives. Not sure if this is the intent?

**Comment by ademar** [comment]
Yeah we renamed that namespace to Suave.Filters.

On Thursday, 17 December 2015, Lev Gorodinski <notifications@github.com>
wrote:

> OK, cool!
>
> I think there is a similar issue with the use of applicatives. I don't
> think functions of type:
>
> HttpContext -> Async<HttpContext option>
>
> are best characterized as applicatives. Not sure if this is the intent?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SuaveIO/suave/issues/361#issuecomment-165583124>.
>


-- 
Ademar Gonzalez
+1-647-891-3606
http://ademar.name
https://github.com/ademar

**Comment by eulerfx** [comment]
Nice, much better name. And sorry, don't mean to be a pedant :)

**Comment by haf** [comment]
Sorry about slowness, has been mad around here, will merge today :)

**Comment by haf** [comment]
Will be fixed by #340
